### ISSUE TITLE
research why `let x=4; . . . . . . let x=8;` is legal

### ISSUE THREAD

**Issue by broesamle** [broesamle/RustWorkshop#5]
``` rust
fn main() {
    let apples = 8;
    let fingers = 3;
    let peanuts = 2;

    println!("There are {} apples, we see {} fingers and how many peanuts? {}, exactly.",
        apples, fingers, peanuts);
    let apples = 12;
    println!("Now, there are {} apples, we see {} fingers and how many peanuts? {}, exactly.",
        apples, fingers, peanuts);

}
```

this compiles and, personally, I don't like the fact that `let apples = 12;` does not care about the former variable declaration.

**Comment by agaviria** [comment]
Hi there!  I ran into your repo.  Perhaps I may add some context as this was something you don't see often in other languages.  Rust has `scope` definitions, these are defined inside a `codeblock` which is any code inside curly braces, see below for breakdown

```
opening block -> {     
    // codeblock
} <- closing block. 
``` 

In your example apples variable is being `shadowed` by the new declaration since its not in `scope`.  If we take your example and assign it in` scope`, it will all make sense:  

```
fn main() {
    let apples = 8;
    let fingers = 3;
    let peanuts = 2;

    println!("There are {} apples, we see {} fingers and how many peanuts? {}, exactly.",
             apples,
             fingers,
             peanuts);
    {
        let apples = 12;
        println!("Now, there are {} apples, we see {} fingers and how many peanuts? {}, exactly.",
                 apples,
                 fingers,
                 peanuts);
    }
    println!("Now, there are {} apples, we see {} fingers and how many peanuts? {}, exactly.",
             apples,
             fingers,
             peanuts);
}
```

Basically any code inside a `codeblock` is first class accessed by the compiler.  Once the code is not in scope the compiler continues to read variables in a procedural fashion, meaning variable `shadowing` is in play.


Reference: [(https://doc.rust-lang.org/book/variable-bindings.html#scope-and-shadowing)]

**Comment by broesamle** [comment]
yes, but why does this compile:
```
fn main() {
    let x = 1;
    println!("x = {}", x);
    let x = 2;
    println!("x = {}", x);
    println!("x = {}", x);
}
```
OUTPUT: 
```
x = 1
x = 2
x = 2
```
but this does not:
```
fn main() {
    let x = 1;
    println!("x = {}", x);
    x = 2;
    println!("x = {}", x);
    println!("x = {}", x);
}
```
with an error regarting (im-)mutability of x:
```
src/main.rs:4:5: 4:10 error: re-assignment of immutable variable `x` [E0384]
src/main.rs:4     x = 2;
                  ^~~~~
. . .
```

(Which is what I would expect for both examples.)

**Comment by agaviria** [comment]
Sure, that is because Rust variables are immutable by default.  Immutable, meaning they are unable to change.  You will find that as Rust being a low level systems language, you will need to be explicit about pretty much everything.  

To make this work just add a `mut` keyword to cast the variable as mutable.

```
fn main() {
    let mut x = 1;
    println!("x = {}", x);
    x = 2;
    println!("x = {}", x);
    println!("x = {}", x);
}
```

[playground](https://is.gd/lDa1i8)

**Comment by broesamle** [comment]
sure... it just does happen in the same block, iy does not warn you about the first x being declared already and, to a newcomer it looks like accidentally changing the value of the same immutable. 

from a 'the compiler helps you to avoid typical human errors' it doesn't help much that it is 'just shadowed'.

**Comment by agaviria** [comment]
Well, I agree that compiler errors are not optimized for readability just yet.  I think that is going to change real soon.  I read this week a posting from Jonathan Turner about getting the community involved to fix this issue.  

Reference: [Helping out with rust errors](http://www.jonathanturner.org/2016/08/helping-out-with-rust-errors.html)

Just to go one step further so any newcomers may read this thread, can understand:  consider the following...

```
fn main() {
    let mut x = 1;
    println!("x = {}", x);
    println!("x = {:p}", &x);
    x = 2;
    println!("x = {}", x);
    println!("x = {:p}", &x);
    {
    let mut x = 1;
    println!("x = {}", x);
    println!("x = {:p}", &x);
    }
}
```

This example would issue a compiler "warning" but it would still compile and run:

```

warning: variable does not need to be mutable, #[warn(unused_mut)] on by default
  --> <anon>:10:9
10 |>     let mut x = 1;
   |>         ^^^^^
-----------------------------------------------------------------------------------
x = 1
x = 0x7fff12739dcc
x = 2
x = 0x7fff12739dcc
x = 1
x = 0x7fff12739ba4

Program ended.
```

The compiler is actually optimized to let the user know that there is no need to mutate at the second codeblock because it actually can infer that it's not being mutated anywhere else in the program.  This in my opinion is fantastic!  In other words, the compiler is yelling at the user, politely....Hey no need to use `mut`. 

Now, what we did here was just use a reference to print out the address of each variable declaration so that you can see that the first two instances of x are pointing to the same address location.  However, the inner codeblock holds a different address regardless of it's mutation.

**Comment by broesamle** [comment]
Thank you for the elaborate reply! (And sorry for me being silent/offline for so long). Regarding beginners workshop material I think it is reasonable to bypass the above complexities. 

I'll have the future error message optimisation on the back of my mind when writing up the workshop/tutorial -- happy to help out with it as soon as I have some time available again (presumably after RustFest).
